src/zope/interface/registry.py

```diff
@@ -38,6 +41,129 @@ def notify(*arg, **kw): pass
 from zope.interface._compat import STRING_TYPES
 
 
+class _UnhashableComponentCounter(object):
+    # defaultdict(int)-like object for unhashable components
+
+    def __init__(self, otherdict):
+        # [(component, count)]
+        self._data = [item for item in otherdict.items()]
+
+    def __getitem__(self, key):
+        for component, count in self._data:
+            if component == key:
+                return count
+        return 0
+
+    def __setitem__(self, component, count):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                self._data[i] = component, count
+                return
+        self._data.append((component, count))
+
+    def __delitem__(self, component):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                del self._data[i]
+                return
+        raise KeyError(component) # pragma: no cover
+
+
+class _UtilityRegistrations(object):
+
+    _regs_for_components = WeakKeyDictionary()
+
+    @classmethod
+    def for_components(cls, components):
+        # We manage these utility/subscription registrations as associated
+        # objects with a weakref to avoid making any changes to
+        # the pickle format
+        try:
+            regs = cls._regs_for_components[components]
+        except KeyError:
+            regs = None
+        else:
+            # In case the components have been re-initted, clear the cache
+            # (zope.component.testing does this between tests)
+            if (regs._utilities is not components.utilities
+                or regs._utility_registrations is not components._utility_registrations):
+                regs = None
```

Comment:
Nope.  Because it doesn't deal with persistent registries.

I have a hack that seems to fix the shoobx tests, but I don't like it.  Basically, if a Components sees it is cached in its __init__, it assumes it's being called from test setUp or tearDown.  This is nasty.

Amongst the things that are bad here:

- __init__ has 2 responsibilities, initialization and test cleanup.  There should be an explicit cleanup mechanism that deals both with the base registry and the cache (if we keep the cache the way it is).

- The complicated weakref lazy cache this PR introduced.  This should be changed to an attribute, with a mechanism to create it for old instances or a _v_ attr.  The latter might be necessary if persistent registrations might somehow be made in the non-persistent registry, but I'm guessing this isn't a problem.